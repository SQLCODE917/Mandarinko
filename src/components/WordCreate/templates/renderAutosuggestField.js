import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Autosuggest from 'react-autosuggest'
import ErrorLine from './errorLine.js'
import styles from './renderAutosuggestField.css'
import theme from './AutosuggestTheme.css'

export default class renderAutosuggestField extends Component {
  constructor(props) {
    super(props)

    this.state = {
      value: '',
      suggestions: []
    }

    this.onChange = this.onChange.bind(this)
    this.onSuggestionSelected = this.onSuggestionSelected.bind(this)
    this.onSuggestionsClearRequested = this.onSuggestionsClearRequested.bind(this)
    this.onSuggestionsFetchRequested = this.onSuggestionsFetchRequested.bind(this)
  }

  // Autosuggest will call this function every time you need to clear suggestions.
  onSuggestionsClearRequested = () => {
    this.setState({
      value: '',
      suggestions: []
    })
  }

  onSuggestionsFetchRequested = ({ value }) => {
    const { getSuggestions } = this.props
    this.setState({
      suggestions: getSuggestions(value)
    })
  }

  onChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    })
  }

  onSuggestionSelected = (event, { suggestion }) => {
    this.setState({
      value: suggestion.id
    })
    const { input } = this.props
    input.onChange(suggestion.id)
  }

  componentWillReceiveProps(nextProps) {
    const { input: { value } } = nextProps
    if (value !== void(0)) {
      this.setState({
        value
      })
    } 
  }
  
  render() {
    const {
      input,
      label,
      type,
      placeholder,
      meta: {
        touched,
        error
      },
      renderSuggestion,
      getSuggestionValue
    } = this.props

    const {
      value,
      suggestions
    } = this.state

    const inputProps = {
      placeholder,
      value,
      onChange: this.onChange,
      type
    }

    return (
      <div className={styles.autosuggestField}>
        <label>{label}</label>
        <div className={styles.autosuggestFieldBody}>
          <Autosuggest
            theme={theme}
            suggestions={suggestions}
            onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
            onSuggestionsClearRequested={this.onSuggestionsClearRequested}
            onSuggestionSelected={this.onSuggestionSelected}
            getSuggestionValue={getSuggestionValue}
            renderSuggestion={renderSuggestion}
            inputProps={inputProps}
          />
        </div>
        { touched && error && <ErrorLine>{error}</ErrorLine>}
      </div>
    )
  }
}

renderAutosuggestField.propTypes = {
  getSuggestions: PropTypes.func.isRequired,
  renderSuggestion: PropTypes.func.isRequired,
  getSuggestionValue: PropTypes.func.isRequired
}
