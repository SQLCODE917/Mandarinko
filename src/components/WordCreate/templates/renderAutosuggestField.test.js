import React from 'react'
import renderAutosuggestField from './renderAutosuggestField.js'
import {
  getSuggestions,
  renderSuggestion,
  getSuggestionValue
} from './ByDefinition.js'
import ErrorLine from './errorLine.js'

import { mount } from 'enzyme'

const allWords = require('../../../../testdata/testVocabulary.json')
const Autosuggest = renderAutosuggestField

describe('Autosuggest Word By Definition', () => {
  let wrapper, getSuggestionsSpy, inputChangeSpy

  beforeEach(() => {
    inputChangeSpy = jest.fn()
    const input = {
      onChange: inputChangeSpy
    }

    getSuggestionsSpy = jest.fn()
      .mockImplementation(getSuggestions(allWords))

    const props = {
      getSuggestions: getSuggestionsSpy,
      renderSuggestion: renderSuggestion,
      getSuggestionValue: getSuggestionValue,
      meta: {},
      type: 'text',
      input
    }
    wrapper = mount(<Autosuggest {...props}/>)
  })

  it('should render', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should trigger autosuggest on input', () => {
    const searchString = 'a'
    const input = wrapper.find('input')
    input.value = searchString
    input.simulate('change', {
      target: {
        value: searchString
      }
    })

    expect(getSuggestionsSpy.mock.calls).toEqual([[searchString]])
  })

  it('should clear out suggestions when requested', () => {
    wrapper.setState({value: 'hello', suggestions: ['world']})
    expect(wrapper.state().value).toEqual('hello')
    const { onSuggestionsClearRequested } = wrapper.instance()
    onSuggestionsClearRequested()
    expect(wrapper.state()).toEqual({ value: '', suggestions: []})
  })

  it('should stimulate the input on suggestion selected', () => {
    const { onSuggestionSelected } = wrapper.instance()
    const suggestion = { id: 'id' }
    onSuggestionSelected('event', { suggestion })
    expect(inputChangeSpy).toHaveBeenCalledWith('id')
  })

  it('should update suggestion when updated in new props', () => {
    expect(wrapper.state().value).toEqual('')
    wrapper.setProps({ input: { value: 'hello' }})
    expect(wrapper.state().value).toEqual('hello')
  })

  it('should not update suggestion when none given in new props', () => {
    expect(wrapper.state().value).toEqual('')
    const { componentWillReceiveProps } = wrapper.instance()
    componentWillReceiveProps({ input: {}})
    expect(wrapper.state().value).toEqual('')
  })

  it('shows errors as they happen', () => {
    expect(wrapper.find(ErrorLine)).toHaveLength(0)
    wrapper.setProps({ meta: { touched: true, error: 'ayyyy' }})
    expect(wrapper.find(ErrorLine)).toHaveLength(1)
  })
})
