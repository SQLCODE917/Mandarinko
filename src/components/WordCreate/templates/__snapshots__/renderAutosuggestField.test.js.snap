// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autosuggest Word By Definition should render 1`] = `
<renderAutosuggestField
  getSuggestionValue={[Function]}
  getSuggestions={[MockFunction]}
  input={
    Object {
      "onChange": [MockFunction],
    }
  }
  meta={Object {}}
  renderSuggestion={[Function]}
  type="text"
>
  <div
    className="autosuggestField"
  >
    <label />
    <div
      className="autosuggestFieldBody"
    >
      <Autosuggest
        alwaysRenderSuggestions={false}
        focusInputOnSuggestionClick={true}
        getSuggestionValue={[Function]}
        highlightFirstSuggestion={false}
        id="1"
        inputProps={
          Object {
            "onChange": [Function],
            "placeholder": undefined,
            "type": "text",
            "value": "",
          }
        }
        multiSection={false}
        onSuggestionSelected={[Function]}
        onSuggestionsClearRequested={[Function]}
        onSuggestionsFetchRequested={[Function]}
        renderSuggestion={[Function]}
        renderSuggestionsContainer={[Function]}
        shouldRenderSuggestions={[Function]}
        suggestions={Array []}
        theme={null}
      >
        <Autowhatever
          getSectionItems={[Function]}
          highlightedItemIndex={null}
          highlightedSectionIndex={null}
          id="1"
          inputProps={
            Object {
              "onBlur": [Function],
              "onChange": [Function],
              "onFocus": [Function],
              "onKeyDown": [Function],
              "placeholder": undefined,
              "type": "text",
              "value": "",
            }
          }
          itemProps={[Function]}
          items={Array []}
          multiSection={false}
          renderInputComponent={[Function]}
          renderItem={[Function]}
          renderItemData={
            Object {
              "query": "",
            }
          }
          renderItemsContainer={[Function]}
          renderSectionTitle={[Function]}
          theme={Object {}}
        >
          <div
            aria-expanded={false}
            aria-haspopup="listbox"
            aria-owns="react-autowhatever-1"
            key="react-autowhatever-1-container"
            role="combobox"
            style={Object {}}
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-controls="react-autowhatever-1"
              autoComplete="off"
              key="react-autowhatever-1-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              style={Object {}}
              type="text"
              value=""
            />
            <div
              id="react-autowhatever-1"
              key="react-autowhatever-1-items-container"
              role="listbox"
              style={Object {}}
            />
          </div>
        </Autowhatever>
      </Autosuggest>
    </div>
  </div>
</renderAutosuggestField>
`;
